feat(forms): introduce new Tiles component

austin94 opened the pull request:
## Description

This PR introduces the new `Tiles` pattern. Semantically this is similar to a [Radio Group](https://www.w3.org/TR/wai-aria-practices/#radiobutton), but this implementation uses native `<input type="radio" />` elements rather than roving tabIndex.

## Detail

### Usage

Consumers can treat this as a controlled or uncontrolled input. It also allows more traditional `<form>` usages if needed.

### API

Common usage:

```jsx
<Tiles name="form-name" onChange={...} value={...}>
  <Tiles.Tile value="item-1">
    <Tiles.Icon>...</Tiles.Icon>
    <Tiles.Label>label</Tiles.Label>
    <Tiles.Description>description</Tiles.Description>
  </Tiles.Tile>
</Tiles>
```

### Layout and Variants

`Tiles` are currently used in a variety of sizes and layouts with only some needing mobile-responsiveness. Rather than bake all of these layouts into the component consumers can use our `<Grid>` component or their own implementations to define the column layouts.

I have included an `isVertical` layout that helps with modal and block usages.

## Preview

![image](https://user-images.githubusercontent.com/4030377/75467371-61b12600-5940-11ea-9fb1-593857d51173.png)

https://austin-next.netlify.com/forms/#tiles

## Checklist

- [ ] :ok_hand: design updates are Garden Designer approved (add the
      designer as a reviewer)
- [x] :globe_with_meridians: Styleguidist demo is up-to-date (`yarn start`)
- [x] :arrow_left: renders as expected with reversed (RTL) direction
- [x] :wheelchair: analyzed via [axe](https://www.deque.com/axe/) and evaluated using VoiceOver
- [x] :guardsman: includes new unit tests
- [x] :memo: tested in Chrome, Firefox, Safari, Edge, and IE11

ginnywood commented:
One other thing: in a vertical Tile, the space between the icon and the text should be 20px. Looks like it is 16 right now.

austin94 commented:
@ginnywood I've addressed the following feedback:

> My specs were wrong on borders for hover and active states

I've updated these values to match radios/checkboxes

> Icons should remain grey, but get progressively darker on hover and active
> Text color should always match icon color

I've brought over the IconButton styling for this.

> Selected Tile has hover and active states too—should match a Radio/Checkbox

These now match Radio/Checkbox

> Disabled default state background should match disabled Text-input/Dropdown
> Disabled selected state background match disabled Checkbox/Radio

I've combined these two styles. Let me know what you think.

> Instead of “vertical”, I propose “stacked”

I think this makes sense considering all row/column layout is handled by the consumer.

I've published the new changes at https://austin-next.netlify.com/forms/#tiles

jzempel commented on packages/forms/examples/tiles.md:
```diff
@@ -0,0 +1,91 @@
+const DecimalIcon = require('@zendeskgarden/svg-icons/src/16/decimal-stroke.svg').default;
+const CalendarIcon = require('@zendeskgarden/svg-icons/src/16/calendar-stroke.svg').default;
+const CreditCardIcon = require('@zendeskgarden/svg-icons/src/16/credit-card-stroke.svg').default;
+const AsterickIcon = require('@zendeskgarden/svg-icons/src/16/asterisk-stroke.svg').default;
```
* `AsteriskIcon` :wink: #rickroll

ginnywood replied:
Yes, I’d initially designed this to be used for checkboxes too. We don’t currently have a use case, but from a design perspective we’re approved and prepared to introduce that concept whenever the engineering effort makes sense.

austin94 replied:
Updated.

jzempel commented on packages/forms/src/styled/tiles/StyledTileLabel.ts:
```diff
@@ -0,0 +1,39 @@
+  font-weight: ${props => props.theme.fontWeights.semibold};
+  /* stylelint-disable-next-line property-no-unknown */
+  margin-${props => (props.theme.rtl ? 'right' : 'left')}: ${props =>
+  props.isStacked && props.theme.space.base * 13}px;
```
Technically, I think the intent here is to represent 20px of space between the icon and the label. So the more accurate calculation would be `space.base * 5 + iconSize`, where the icon size is `iconSizes.md * 2`.

austin94 replied:
Updated.

jzempel commented on packages/forms/src/styled/tiles/StyledTileIcon.ts:
```diff
@@ -0,0 +1,47 @@
+  top: ${props => props.theme.space.base * 6}px;
+  text-align: center;
+  line-height: 0;
+  ${props => (props.theme.rtl ? 'right' : 'left')}: ${props => props.theme.space.base * 5}px;
```
Neither `right/left` or `top` is applicable for non-stacked tiles. Feels like the absolute positioning could all be moved to `sizeStyles` and simplified by using local variables.

austin94 replied:
I've moved this logic to the existing `sizeStyles` utility to help clarify.

jzempel commented on packages/forms/src/styled/tiles/StyledTileDescription.ts:
```diff
@@ -0,0 +1,35 @@
+  font-size: ${props => props.theme.fontSizes.sm};
+  /* stylelint-disable-next-line property-no-unknown */
+  margin-${props => (props.theme.rtl ? 'right' : 'left')}: ${props =>
+  props.isStacked && props.theme.space.base * 13}px;
```
Same comment here re: the `iconSize` + `margin` calculation.

austin94 replied:
Updated.

jzempel commented on packages/forms/src/styled/tiles/StyledTileDescription.ts:
```diff
@@ -0,0 +1,35 @@
+  display: block;
+  margin-top: ${props => props.theme.space.base}px;
+  text-align: ${props => !props.isStacked && 'center'};
+  line-height: ${props => props.theme.space.base * 4}px;
```
Should be using the `getLineHeight` utility here.

austin94 replied:
Updated.

jzempel commented on packages/forms/src/styled/tiles/StyledTileLabel.ts:
```diff
@@ -0,0 +1,39 @@
+  overflow: hidden;
+  text-align: ${props => !props.isStacked && 'center'};
+  text-overflow: ellipsis;
+  line-height: ${props => props.theme.space.base * 5}px;
```
Should be using the `getLineHeight` utility here.

austin94 replied:
Updated.

jzempel commented on packages/forms/src/styled/tiles/StyledTile.ts:
```diff
@@ -0,0 +1,145 @@
+
+StyledTileInput.defaultProps = {
+  theme: DEFAULT_THEME
+};
```
Would prefer to see this in its own file with the rules for view components applied:
- `data-garden-id` & `data-garden-package` attrs
- theming override capabilities via `retrieveComponentStyles`

austin94 replied:
This is a hidden input that is not visually show and only included for a11y/interaction. I'm not sure why this would need to be themed?

jzempel replied:
I still think the clearer pattern is to separate styled components, and to not assert whether there is theming benefit or not. But YMMV – I'll go with your preference here.

austin94 replied:
I've moved this to a standalone file to separate them.

jzempel commented on packages/forms/src/elements/tiles/components/Tile.tsx:
```diff
@@ -0,0 +1,83 @@
+
+          /**
+           * The focus-visible attribute provided by `react-theming` is
+           * unable to persit to the `<label>`. This manually applies focus
```
*persist

austin94 replied:
Updated.

jzempel commented on packages/forms/src/elements/tiles/components/Tile.tsx:
```diff
@@ -0,0 +1,83 @@
+interface ITileProps {
+  /** The value of the input */
+  value?: string;
+  /** Whether the radio is disabled */
```
Should "radio" be "tile"?

austin94 replied:
Updated.

jzempel commented on packages/forms/src/elements/tiles/Tiles.tsx:
```diff
@@ -0,0 +1,91 @@
+
+/* eslint-disable react/display-name */
+/**
+ * Accepts all `<div>` attributes
```
Without improved documentation here, it is very hard to understand how to use the `Tiles` component just by viewing the API. Unless there is a way to make Styleguidist pick up the `Tiles.Tile` etc. components below, there should probably be a minimal usage example along with showing how it should be incorporated with the `Grid`.

austin94 replied:
I've added a new basic example and some documentation regarding layout and the static property components.

jzempel commented on packages/forms/examples/tiles.md:
```diff
@@ -0,0 +1,164 @@
+  - [`value`] The value of the `input`
+  - [`disabled`] Whether the `Tile` is disabled
+- `Tiles.Label` Accepts all `span` attributes
+- `Tiles.Description` Accepts all `span` attributes
```
What about `Tiles.Icon`?

austin94 replied:
Added!

austin94 commented on packages/forms/src/elements/tiles/Tiles.spec.tsx:
```diff
@@ -0,0 +1,253 @@
+        </Tiles>
+      );
+
+      expect(container.querySelector('input')).toBeChecked();
```
I had just assumed this didn't work for inputs that didn't have a defined `aria-labelledby`! Updated.

austin94 commented on packages/forms/src/elements/tiles/Tiles.spec.tsx:
```diff
@@ -0,0 +1,253 @@
+      expect(container.querySelector('input')).toBeDisabled();
+    });
+
+    it('checks input when tile is selected', () => {
```
I've added an uncontrolled usage test.

hzhu commented on packages/forms/src/elements/tiles/Tiles.spec.tsx:
```diff
@@ -57,47 +57,69 @@ describe('Tiles', () => {
+        </Tiles>
+      );
+
+      fireEvent.click(getByText('label'));
```
Not a blocker, but this should probably use `getByLabelText` also.

austin94 replied:
This would actually the backing `input`. This tests that the `label` is able to be clicked.
